src/MongoDB.Driver.Core/Core/Configuration/ClusterBuilder.cs

```diff
@@ -248,5 +209,79 @@ public ClusterBuilder Subscribe(IEventSubscriber subscriber)
             _eventAggregator.Subscribe(subscriber);
             return this;
         }
+
+        // private methods
+        private IClusterFactory CreateClusterFactory()
+        {
+            var serverFactory = CreateServerFactory();
+
+            return new ClusterFactory(
+                _clusterSettings,
+                serverFactory,
+                _eventAggregator);
+        }
+
+        private IConnectionPoolFactory CreateConnectionPoolFactory()
+        {
+            var streamFactory = CreateTcpStreamFactory(_tcpStreamSettings);
+
+            var connectionFactory = new BinaryConnectionFactory(
+                _connectionSettings,
+                streamFactory,
+                _eventAggregator);
+
+            return new ExclusiveConnectionPoolFactory(
+                _connectionPoolSettings,
+                connectionFactory,
+                _eventAggregator);
+        }
+
+        private ServerFactory CreateServerFactory()
+        {
+            var connectionPoolFactory = CreateConnectionPoolFactory();
+            var serverMonitorFactory = CreateServerMonitorFactory();
+
+            return new ServerFactory(
+                _clusterSettings.ConnectionMode,
+                _serverSettings,
+                connectionPoolFactory,
+                serverMonitorFactory,
+                _eventAggregator);
+        }
+
+        private IServerMonitorFactory CreateServerMonitorFactory()
+        {
+            var serverMonitorConnectionSettings = _connectionSettings
+                .With(authenticators: new IAuthenticator[] { });
+
+            var serverMonitorTcpStreamSettings = new TcpStreamSettings(_tcpStreamSettings)
+                .With(
+                    readTimeout: _serverSettings.HeartbeatTimeout,
+                    writeTimeout: _serverSettings.HeartbeatTimeout
+                );
```

Comment:
The serverMonitorTcpStreamSettings variable is new. It is a copy of the _tcpStreamSettings but with the read/write timeout overridden.

Reply:
The default value for ConnectTimeout is 30 seconds. It should only be Infinite if the user configured it to be Infinite.

If the user explicitly configures it to be Infinite do you think we should override that?

Reply:
Yes, I think we should.  Having any infinite timeouts for the monitoring socket defeats the purpose of monitoring, so let's just not do it.  

There are also valid reasons why someone might set connectTimeout to 0 (infinite) as the OS typically has its own default connect timeout that an app may not wish to override. 

I'm going to open a bug for Java driver about that as well, as it doesn't do that.

Reply:
Done.